server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java

```diff
@@ -296,15 +296,129 @@ protected LeafBucketCollector getLeafCollector(AggregationExecutionContext aggCt
             if (QueryToFilterAdapter.matchesNoDocs(filters()) && otherBucketKey == null) {
                 return LeafBucketCollector.NO_OP_COLLECTOR;
             }
+            if (filters().size() == 1) {
+                return new SingleFilterLeafCollector(
+                    sub,
+                    aggCtx.getLeafReaderContext(),
+                    filters().get(0),
+                    totalNumKeys,
+                    parent != null,
+                    otherBucketKey != null
+                );
+            }
+            return new MultiFilterLeafCollector(
+                sub,
+                aggCtx.getLeafReaderContext(),
+                filters(),
+                totalNumKeys,
+                parent != null,
+                otherBucketKey != null
+            );
+        }
+    }
+
+    private abstract static class AbstractLeafCollector extends LeafBucketCollectorBase {
+        final LeafBucketCollector sub;
+        final int totalNumKeys;
+        final boolean hasOtherBucket;
+
+        AbstractLeafCollector(LeafBucketCollector sub, int totalNumKeys, boolean hasOtherBucket) {
+            super(sub, null);
+            this.sub = sub;
+            this.totalNumKeys = totalNumKeys;
+            this.hasOtherBucket = hasOtherBucket;
+        }
+
+        final long bucketOrd(long owningBucketOrdinal, int filterOrd) {
+            return owningBucketOrdinal * totalNumKeys + filterOrd;
+        }
+    }
+
+    private class SingleFilterLeafCollector extends AbstractLeafCollector {
+
+        final FilterMatchingDisiWrapper filterIterator;
+        final QueryToFilterAdapter filter;
+        final boolean usesCompetitiveIterator;
+
+        SingleFilterLeafCollector(
+            LeafBucketCollector sub,
+            LeafReaderContext leafReaderContext,
+            QueryToFilterAdapter filter,
```

Comment:
I wonder if we should convert `QueryToFilterAdapter` objects to `Scorer`s up-front. This way we could also use the `Single*` implementation if there are multiple filters but only one of them has matches on the segment, and we could use the `Multi*` implementation when no filters have matches on the segments, which would in-turn help remove the `filterIterator != null` check in `SingleFilterLeafCollector#collect`?

Reply:
Yeah, this indeed can improve performance in more cases, while the code is cleaner since more logic is moved outside the subclasses.